File: libtest/src/main/java/z/util/primitives/Primitives.java
/**
 * Copyright 2013, Landz and its contributors. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */



package z.util.primitives;

/*
 * #%L
 * libtest
 * %%
 * Copyright (C) 2015 - 2017 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * modified by:
 * @auther Landz's contributors
 */

/**
 * NOTE:
 *   Partially form APLed Guava library, but evolves in the Landz independently.
 * Contains static utility methods pertaining to primitive types and their
 * corresponding wrapper types.
 *
 * @author Kevin Bourrillion
 */
public final class Primitives {
  private Primitives() {}

  /** A map from primitive types to their corresponding wrapper types. */
  private static final Map<Class<?>, Class<?>> PRIMITIVE_TO_WRAPPER_TYPE;

  /** A map from wrapper types to their corresponding primitive types. */
  private static final Map<Class<?>, Class<?>> WRAPPER_TO_PRIMITIVE_TYPE;

  // Sad that we can't use a BiMap. :(
  
  static {
    Map<Class<?>, Class<?>> primToWrap = new HashMap<Class<?>, Class<?>>(16);
    Map<Class<?>, Class<?>> wrapToPrim = new HashMap<Class<?>, Class<?>>(16);

    add(primToWrap, wrapToPrim, boolean.class, Boolean.class);
    add(primToWrap, wrapToPrim, byte.class, Byte.class);
    add(primToWrap, wrapToPrim, char.class, Character.class);
    add(primToWrap, wrapToPrim, double.class, Double.class);
    add(primToWrap, wrapToPrim, float.class, Float.class);
    add(primToWrap, wrapToPrim, int.class, Integer.class);
    add(primToWrap, wrapToPrim, long.class, Long.class);
    add(primToWrap, wrapToPrim, short.class, Short.class);
    add(primToWrap, wrapToPrim, void.class, Void.class);

    PRIMITIVE_TO_WRAPPER_TYPE = Collections.unmodifiableMap(primToWrap);
    WRAPPER_TO_PRIMITIVE_TYPE = Collections.unmodifiableMap(wrapToPrim);
  }

  private static void add(Map<Class<?>, Class<?>> forward,
      Map<Class<?>, Class<?>> backward, Class<?> key, Class<?> value) {
    forward.put(key, value);
    backward.put(value, key);
  }

  /**
   * Returns an immutable set of all nine primitive types (including {@code
   * void}). Note that a simpler way to test whether a {@code Class} instance
   * is a member of this set is to call {@link Class#isPrimitive}.
   */
  public static Set<Class<?>> allPrimitiveTypes() {
    return PRIMITIVE_TO_WRAPPER_TYPE.keySet();
  }

  /**
   * Returns an immutable set of all nine primitive-wrapper types (including
   * {@link Void}).
   */
  public static Set<Class<?>> allWrapperTypes() {
    return WRAPPER_TO_PRIMITIVE_TYPE.keySet();
  }

  /**
   * Returns {@code true} if {@code type} is one of the nine
   * primitive-wrapper types, such as {@link Integer}.
   *
   * @see Class#isPrimitive
   */
  public static boolean isWrapperType(Class<?> type) {
    return WRAPPER_TO_PRIMITIVE_TYPE.containsKey(type);
  }

  /**
   * Returns the corresponding wrapper type of {@code type} if it is a primitive
   * type; otherwise returns {@code type} itself. Idempotent.
   * <pre>
   *     wrap(int.class) == Integer.class
   *     wrap(Integer.class) == Integer.class
   *     wrap(String.class) == String.class
   * </pre>
   */
  public static <T> Class<T> wrap(Class<T> type) {
    // cast is safe: long.class and Long.class are both of type Class<Long>
    @SuppressWarnings("unchecked")
    Class<T> wrapped = (Class<T>) PRIMITIVE_TO_WRAPPER_TYPE.get(type);
    return (wrapped == null) ? type : wrapped;
  }

  /**
   * Returns the corresponding primitive type of {@code type} if it is a
   * wrapper type; otherwise returns {@code type} itself. Idempotent.
   * <pre>
   *     unwrap(Integer.class) == int.class
   *     unwrap(int.class) == int.class
   *     unwrap(String.class) == String.class
   * </pre>
   */
  public static <T> Class<T> unwrap(Class<T> type) {
    // cast is safe: long.class and Long.class are both of type Class<Long>
    @SuppressWarnings("unchecked")
    Class<T> unwrapped = (Class<T>) WRAPPER_TO_PRIMITIVE_TYPE.get(type);
    return (unwrapped == null) ? type : unwrapped;
  }
}

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/nvm/primitives.kt
/*-
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2019 - 2020 JetBrains s.r.o.
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */
package org.jetbrains.kotlinx.lincheck.nvm

import kotlinx.atomicfu.atomic
import org.jetbrains.kotlinx.lincheck.runner.RecoverableStateContainer


private fun thread() = RecoverableStateContainer.threadId()

abstract class AbstractNonVolatilePrimitive {
    protected val empty = BooleanArray(NVMCache.MAX_THREADS_NUMBER) { true }

    abstract fun flushInternal(threadId: Int = thread())
    fun flush(threadId: Int = thread()) {
        flushInternal(threadId)
        NVMCache.remove(threadId, this)
    }

    internal fun crash(threadId: Int) {
        if (Probability.shouldFlush()) {
            flushInternal(threadId)
        }
        empty[threadId] = true
    }
}

fun nonVolatile(value: Int) = NonVolatileInt(value)
fun nonVolatile(value: Long) = NonVolatileLong(value)
fun nonVolatile(value: Boolean) = NonVolatileBoolean(value)
fun <T> nonVolatile(value: T) = NonVolatileRef(value)

/** Persistent reference emulates non-volatile memory variable with volatile cache. */
class NonVolatileRef<T> internal constructor(initialValue: T) : AbstractNonVolatilePrimitive() {
    @Volatile
    private var nonVolatileValue = initialValue
    private val volatileValue = atomic(initialValue)

    fun read(threadId: Int = thread()) = if (empty[threadId]) nonVolatileValue else volatileValue.value

    fun write(value: T, threadId: Int = thread()) {
        empty[threadId] = false
        volatileValue.value = value
        NVMCache.add(threadId, this)
    }

    override fun flushInternal(threadId: Int) {
        if (empty[threadId]) return
        nonVolatileValue = volatileValue.value
    }

    fun writeAndFlush(value: T, threadId: Int = RecoverableStateContainer.threadId()) {
        empty[threadId] = false
        volatileValue.value = value
        nonVolatileValue = value
    }

    fun compareAndSet(expect: T, update: T, threadId: Int = RecoverableStateContainer.threadId()): Boolean {
        if (empty[threadId]) {
            volatileValue.value = nonVolatileValue
        }
        NVMCache.add(threadId, this)
        return volatileValue.compareAndSet(expect, update)
    }

    fun getAndSet(value: T, threadId: Int = RecoverableStateContainer.threadId()): T {
        if (empty[threadId]) {
            volatileValue.value = nonVolatileValue
        }
        NVMCache.add(threadId, this)
        return volatileValue.getAndSet(value)
    }
}

class NonVolatileInt internal constructor(initialValue: Int) : AbstractNonVolatilePrimitive() {
    @Volatile
    private var nonVolatileValue = initialValue
    private val volatileValue = atomic(initialValue)

    fun read(threadId: Int = thread()) = if (empty[threadId]) nonVolatileValue else volatileValue.value

    fun write(value: Int, threadId: Int = thread()) {
        empty[threadId] = false
        volatileValue.value = value
        NVMCache.add(threadId, this)
    }

    override fun flushInternal(threadId: Int) {
        if (empty[threadId]) return
        nonVolatileValue = volatileValue.value
    }

    fun writeAndFlush(value: Int, threadId: Int = RecoverableStateContainer.threadId()) {
        empty[threadId] = false
        volatileValue.value = value
        nonVolatileValue = value
    }

    fun compareAndSet(expect: Int, update: Int, threadId: Int = RecoverableStateContainer.threadId()): Boolean {
        if (empty[threadId]) {
            volatileValue.value = nonVolatileValue
        }
        NVMCache.add(threadId, this)
        return volatileValue.compareAndSet(expect, update)
    }

    fun getAndSet(value: Int, threadId: Int = RecoverableStateContainer.threadId()): Int {
        if (empty[threadId]) {
            volatileValue.value = nonVolatileValue
        }
        NVMCache.add(threadId, this)
        return volatileValue.getAndSet(value)
    }
}


class NonVolatileLong internal constructor(initialValue: Long) : AbstractNonVolatilePrimitive() {
    @Volatile
    private var nonVolatileValue = initialValue
    private val volatileValue = atomic(initialValue)

    fun read(threadId: Int = thread()) = if (empty[threadId]) nonVolatileValue else volatileValue.value

    fun write(value: Long, threadId: Int = thread()) {
        empty[threadId] = false
        volatileValue.value = value
        NVMCache.add(threadId, this)
    }

    override fun flushInternal(threadId: Int) {
        if (empty[threadId]) return
        nonVolatileValue = volatileValue.value
    }

    fun writeAndFlush(value: Long, threadId: Int = RecoverableStateContainer.threadId()) {
        empty[threadId] = false
        volatileValue.value = value
        nonVolatileValue = value
    }

    fun compareAndSet(expect: Long, update: Long, threadId: Int = RecoverableStateContainer.threadId()): Boolean {
        if (empty[threadId]) {
            volatileValue.value = nonVolatileValue
        }
        NVMCache.add(threadId, this)
        return volatileValue.compareAndSet(expect, update)
    }

    fun getAndSet(value: Long, threadId: Int = RecoverableStateContainer.threadId()): Long {
        if (empty[threadId]) {
            volatileValue.value = nonVolatileValue
        }
        NVMCache.add(threadId, this)
        return volatileValue.getAndSet(value)
    }
}

class NonVolatileBoolean internal constructor(initialValue: Boolean) : AbstractNonVolatilePrimitive() {
    @Volatile
    private var nonVolatileValue = initialValue
    private val volatileValue = atomic(initialValue)

    fun read(threadId: Int = thread()) = if (empty[threadId]) nonVolatileValue else volatileValue.value

    fun write(value: Boolean, threadId: Int = thread()) {
        empty[threadId] = false
        volatileValue.value = value
        NVMCache.add(threadId, this)
    }

    override fun flushInternal(threadId: Int) {
        if (empty[threadId]) return
        nonVolatileValue = volatileValue.value
    }

    fun writeAndFlush(value: Boolean, threadId: Int = RecoverableStateContainer.threadId()) {
        empty[threadId] = false
        volatileValue.value = value
        nonVolatileValue = value
    }

    fun compareAndSet(expect: Boolean, update: Boolean, threadId: Int = RecoverableStateContainer.threadId()): Boolean {
        if (empty[threadId]) {
            volatileValue.value = nonVolatileValue
        }
        NVMCache.add(threadId, this)
        return volatileValue.compareAndSet(expect, update)
    }

    fun getAndSet(value: Boolean, threadId: Int = RecoverableStateContainer.threadId()): Boolean {
        if (empty[threadId]) {
            volatileValue.value = nonVolatileValue
        }
        NVMCache.add(threadId, this)
        return volatileValue.getAndSet(value)
    }
}

